<launch>

  <arg name="robot" default="orcvio_robot"/>
  <arg name="fixed_frame_id" default="world"/>

  <!-- bag parameters -->
  <arg name="path_bag" default="" />
  
  <arg name="visualize_rviz" default="true" />
  
  <!-- evaluation mode is used to compare the results of different flags -->
  <arg name="evaluation_mode" default="false" />

  <arg name="image_topic" default="/camera/image_raw" />

  <node name="republish_img" type="republish" pkg="image_transport" output="screen" args="compressed in:=/camera/infra1/image_rect_raw  raw out:=$(arg image_topic)" />

  <node name="republish_darknet_img" type="republish" pkg="image_transport" output="screen" args="compressed in:=/darknet_ros/detection_image  raw out:=/darknet_ros/detection_image" />

  <!-- System Manager Nodelet  -->
  <group ns="$(arg robot)">
    <node pkg="nodelet" type="nodelet" name="system"
      args='standalone orcvio/SystemNodelet'
      output="screen">

      <param name="fixed_frame_id" value="$(arg fixed_frame_id)"/>

      <remap from="~imu" to="/camera/imu"/>
      <remap from="~cam0_image" to="$(arg image_topic)"/>

      <param name="config_file" value="$(find orcvio)/../../../config/rs_d435i.yaml"/>
      <param name="result_file" value="$(find orcvio)/../../../cache/temp_rmse.txt"/>

      <param name="load_groundtruth_flag" type="int" value="0"/>
      <!-- <param name="use_left_perturbation_flag" type="int" value="1" /> -->

      <param name="publish_tf" type="bool" value="true" />

    </node>

    <!-- Object Mapper Nodelet  -->
    <!-- <node pkg="nodelet" type="nodelet" name="ObjectMapper"
      args="standalone orcvio/ObjectMapperNodelet" > -->
    
    <!-- to see output in terminal  -->
    <node pkg="nodelet" type="nodelet" name="ObjectMapper"
      args="standalone orcvio/ObjectMapperNodelet"
      output="screen">

    <!-- to use gdb  -->
    <!-- <node pkg="nodelet" type="nodelet" name="ObjectMapper"
      args="standalone orcvio/ObjectMapperNodelet"
      output="screen"
      launch-prefix="xterm -e gdb dash-args"
      > -->

          <param name="fixed_frame_id" value="$(arg fixed_frame_id)"/>

          <!-- camera extrinsics -->
          <rosparam param="T_cam_imu">
              [
              0.9999662089030257, -0.006906390871642436, -0.004459015276860498, 0.008708247801839689, 
              0.006874866759746373, 0.9999515383335019, -0.007046785898527633, -0.02174943710377053, 
              0.004507467043353249, 0.007015892643993886, 0.9999652293911333, -0.022785325544131582, 
              0., 0., 0., 1. 
              ]
          </rosparam>

          <!-- bag topics -->
          <param name="topic_bbox"  type="string" value="/sort_ros/tracked_bounding_boxes" />

          <param name="topic_image"  type="string" value="$(arg image_topic)" />
          <param name="topic_caminfo"  type="string" value="/camera/infra1/camera_info" />          
          <param name="topic_pose"  type="string" value="/orcvio_robot/system/poseout" />

          <param name="max_object_feature_track"             type="int"    value="20" />
          <param name="min_object_feature_track"               type="int"    value="5" />

          <!-- object feature -->
          <rosparam command="load" file="$(find orcvio)/../../../config/object_feat_erl.yaml"/>
          <param name="load_gt_object_info_flag"  type="bool" value="false" />

          <param name="to_color_image_flag"  type="bool" value="true" />
          
          <param name="use_unity_dataset_flag"  type="bool" value="false" />


          <!-- this has to be consistent with VIO config  -->
          <param name="use_left_perturbation_flag"  type="bool" value="false" />
          <param name="use_new_bbox_residual_flag"  type="bool" value="true" />
          
    </node>
  </group>

    <!-- play the dataset -->
    <node pkg="rosbag" type="play" name="rosbag" args="$(arg path_bag)" required="true"/>
    
    <!-- open rviz  -->
    <group if="$(arg visualize_rviz)" >
      <node type="rviz" name="rviz" pkg="rviz" args="-d $(find orcvio)/../../../ros_wrapper/src/orcvio/launch/rviz/orcvio_mapping_rs_d435i.rviz" />
    </group>

</launch>
